ambry-commons/src/main/java/com/github/ambry/commons/CommonUtils.java

```diff
@@ -15,50 +15,60 @@
+import com.github.ambry.utils.Pair;
 import java.util.List;
-import org.apache.helix.manager.zk.ZNRecordSerializer;
 import org.apache.helix.manager.zk.ZkBaseDataAccessor;
-import org.apache.helix.manager.zk.ZkClient;
 import org.apache.helix.store.HelixPropertyStore;
 import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.apache.helix.zookeeper.api.client.RealmAwareZkClient;
 import org.apache.helix.zookeeper.api.client.ZkClientType;
 import org.apache.helix.zookeeper.datamodel.ZNRecord;
+import org.apache.helix.zookeeper.datamodel.serializer.ZNRecordSerializer;
 
 
 public class CommonUtils {
   /**
-   * Create a instance of {@link HelixPropertyStore} based on given {@link HelixPropertyStoreConfig}.
+   * Create an instance of {@link HelixPropertyStore} and {@link ZkBaseDataAccessor} based on given {@link HelixPropertyStoreConfig}.
+   * Close {@link ZkBaseDataAccessor} after stopping {@link HelixPropertyStore}.
    * @param zkServers the ZooKeeper server address.
    * @param propertyStoreConfig the config for {@link HelixPropertyStore}.
    * @param subscribedPaths a list of paths to which the PropertyStore subscribes.
-   * @return the instance of {@link HelixPropertyStore}.
+   * @return A pair of an instance of {@link HelixPropertyStore} and an instance of {@link ZkBaseDataAccessor}.
    */
-  public static HelixPropertyStore<ZNRecord> createHelixPropertyStore(String zkServers,
-      HelixPropertyStoreConfig propertyStoreConfig, List<String> subscribedPaths) {
+  public static Pair<HelixPropertyStore<ZNRecord>, ZkBaseDataAccessor<ZNRecord>> createHelixPropertyStore(
+      String zkServers, HelixPropertyStoreConfig propertyStoreConfig, List<String> subscribedPaths) {
     if (zkServers == null || zkServers.isEmpty() || propertyStoreConfig == null) {
       throw new IllegalArgumentException("Invalid arguments, cannot create HelixPropertyStore");
     }
-    ZkClient zkClient = new ZkClient(zkServers, propertyStoreConfig.zkClientSessionTimeoutMs,
-        propertyStoreConfig.zkClientConnectionTimeoutMs, new ZNRecordSerializer());
-    return new ZkHelixPropertyStore<>(new ZkBaseDataAccessor<>(zkClient), propertyStoreConfig.rootPath,
-        subscribedPaths);
+    ZkBaseDataAccessor<ZNRecord> baseDataAccessor =
+        new ZkBaseDataAccessor.Builder<ZNRecord>().setZkClientType(ZkClientType.DEDICATED)
+            .setZkAddress(zkServers)
+            .setRealmAwareZkClientConfig(new RealmAwareZkClient.RealmAwareZkClientConfig().setConnectInitTimeout(
+                propertyStoreConfig.zkClientConnectionTimeoutMs).setZkSerializer(new ZNRecordSerializer()))
+            .setRealmAwareZkConnectionConfig(
+                new RealmAwareZkClient.RealmAwareZkConnectionConfig.Builder().setSessionTimeout(
+                    propertyStoreConfig.zkClientSessionTimeoutMs).build())
+            .build();
+    return new Pair<>(new ZkHelixPropertyStore<>(baseDataAccessor, propertyStoreConfig.rootPath, subscribedPaths),
```

Comment:
QQ: Should helix be responsible for closing the baseDataAccessor inside their stop() method since they have baseDataAccessor in HelixPropertyStore?

Reply:
I've tried a couple different ways of fixing this, unfortunately, this seems to be the only way out.

The reason why ZkHelixPropertyStore doesn't close the zookeeper client that we create in the same method, is that ZkHelixPropertyStore doesn't think it owns zookeeper client. We are passing a BaseDataAccessor to ZkHelixPropertyStore, ZkHelixPropertyStore thinks BaseDataAccessor owns the zookeeper client, so it will let BaseDataAccessor close the zookeeper client.

Well, can we make ZkHelixPropertyStore create zookeeper client internally so it would own the zookeeper client? Unfortunately, we can't. Through there are 3 more constructors for ZkHelixPropertyStore that would let it create a zookeeper client internally, none of those 3 constructors meets our requirements. They either fail to let us specify ZkClientType, or watched paths. Unless we use a shared zookeeper client, or we don't cache those znodes, we can't use current constructors. If helix can expose more constructors to enable more flexible object creation, that would solve the issue.

Then, can we just use ZkCacheDataAccessor? It seems to implement HelixPropertyStore and has a rich constructor set. Unfortunately, we can't. Because some HelixManager, as a common interface, is expose method to return concrete class ZkHelixPropertyStore instead of a more general interface HelixPropertyStore. That locks us with ZkHelixPropertyStore instead of more general interface HelixPropertyStore.